DESIGN.md

```diff
@@ -0,0 +1,77 @@
+# Design document
+
+Felek runs arbitrary Unix processes ("jobs") on the server machine, allowing a client to query job status, stop running jobs, and inspect its stdout and stderr streams ("logs").
+
+## Architecture
+
+Felek uses Go's `exec.Cmd` to run external processes, and provides a gRPC interface for clients.
+
+When a `Start` request is received, Felek will try to start an external process using the provided settings (command path, arguments, environment variables, and a working directory path). If the process was started successfully, Felek will generate a unique ID for it an store it in a key-value store for later queries.
+
+The key-value store for started jobs is protected by an `RWMutex` to allow for many readers or a single writer. `Start` and `Stop` require a write lock, while other RPCs require a read lock.
+
+For each created job, Felek creates two files to store the process's stdout and stderr streams, respectively. This allows easily querying existing logs as well as "following" logs for streaming, for example using existing wrapper around `inotify` such as [`hpcloud/tail`](https://github.com/hpcloud/tail). Felek starts a goroutine for each job to `.Wait()` on the job to complete and then close the file handles to its log files.
+
+For querying logs, Felek use gRPC's streaming functionality to stream logs line-by-line, optionally continuing to stream after the existing content of the log file is exhaust. Logs are assumed to consist of lines of UTF-8 encoded text.
+
+Connections from the client to the server will be encrypted and authenticated using mTLS.
+
+## mTLS setup
+
+### Authentication
+
+We create a self-signed root X.509 CA  and an intermediate CA signed by the root.
```

Comment:
This is good, could you please add more information about how TLS will be configured.

Reply:
Do you mean the specifics of the generated keys and certificates, or the server configuration, or both?

Reply:
Yeah, I meant server TLS configuration.